### PR TITLE
(Nested)Integral of (Nested)summations works. fixes #7827

### LINKED ISSUE #7827: Integration of summations:
I want to compute an integral like (`n` and `N` are integers)
`integrate(summation(sin(n*x), (n,1,N)), x)`
but SymPy returns the same, i.e., it cannot interchange the summation and the integral.

For derivatives it works, though
`diff(summation(sin(n*x), (n,1,N)), x)`
returns
`Sum(n*cos(n*x), (n,1,N))` .

This feature can be useful, e.g.,  when solving differential equations with harmonic forcing.

### PR DESCRIPTION
**Pull request by kaushik94**
This implements integral of summations. Nested integrals and summations are also fixed. Please take a look at the tests. @asmeurer please review. fixes #7827

```
from sympy import *
x, n, N, y = symbols('x n N y')
print integrate(summation(sin(n*x), (n,1,N)), x)
print integrate(Sum(x, (x, 1, y)), x)
print integrate(Sum(x, (x, y, n)), y)
print integrate(Sum(x, (x, n, y)), y)
print integrate(Sum(x, (x, 1, n)), (x, n))
print integrate(Sum(x, (x, y, n)), (x, n, y))
```

output :

```
Sum(Piecewise((0, n == 0), (-cos(n*x)/n, True)), (n, 1, N))
x*Sum(x, (x, 1, y))
Integral(Sum(x, (x, y, n)), y)
Integral(Sum(x, (x, n, y)), y)
n*Sum(x, (x, 1, n))
-n*Sum(x, (x, y, n)) + y*Sum(x, (x, y, n))
```

### CODE REVIEW

**Comment by kaushik94** [comment]
#7827 , #8771

**Comment by smichr** [comment]
In the previous comment you wrote issue/PR numbers which is nice (because they create links) but they would be more useful in the first text box of this page (called the OP, opening paragraph, or description of the PR). 

Also, if you follow [this syntax](https://help.github.com/articles/closing-issues-via-commit-messages/) then the issues/PR will be closed when this is committed. (That syntax, however, must appear in the OP, not in the discussion area where you have put it.) I do see that you included the "fixes #7827" in the commit message -- that will close it, too -- but it is all around more useful in the OP of the PR.

File: sympy/integrals/integrals.py [hunk 1]
```diff
@@ -394,6 +395,8 @@ def doit(self, **hints):
             return S.Zero
 
         # now compute and check the function
+        if isinstance(self.function, Sum):
+            return Sum(integrate(self.function.function, self.variables[0]), self.function.limits)
```
Comment (smichr) [inline sympy/integrals/integrals.py]: On #7827 @asmeurer mentioned that one should "go inside summations that don't have the integration variable in the limits". You haven't checked for that here. e.g., perhaps he was thinking of something like this: `integrate(Sum(x, (x, 1, y)), y)`:

```python
>>> self = Integral(Sum(x,(x,1,y)),y)
>>> Sum(integrate(self.function.function, self.variables[0]),
... self.function.limits)
Sum(x*y, (x, 1, y))
>>> _.doit().expand()
y**3/2 + y**2/2
```

but compare to 

```python
>>> self.doit().expand()
y**3/6 + y**2/4
```

Comment (kaushik94) [inline sympy/integrals/integrals.py]: Right. Will fix that @smichr

**Comment by kaushik94** [comment]
@smichr  added check on limits, added proper syntax in message box. Please take a look

**Comment by smichr** [comment]
You only considered one limit. We might have an integration as 

- integrate(s, x)
- integrate(s, (x,a))
- integrate(s, (x, a, b))

Summations only support limits of length 3, I think, so we have to consider under what conditions s = Sum(f, (y, c, d)) will not allow the integral to be put inside. I don't know the answer, but that is the question that should be clearly answered before this is committed. It would be good to put the answer in the OP at the top to show others that this is ready to go.

**Comment by kaushik94** [comment]
@smichr  please take a look now. Im not sure if this is what you are expecting ?. And its showing merge conflicts, what went wrong ?

**Comment by smichr** [comment]
> what went wrong

master changed; update your master and then merge this branch with it (see "Synchornization" [here](https://github.com/sympy/sympy/wiki/Development-workflow)).

**Comment by smichr** [comment]
These are the cases in master where an Integral is returned

```
>>> s
Sum(a + x, (x, b, c))
>>> from sympy.utilities.iterables import permutations, subsets
>>> for l in subsets((x, a, b, c)):
...   if not l: continue
...   for p in permutations(l):
...    if isinstance(integrate(s, *p), Integral):print p
...
(a,)
(b,)
(c,)
(a, b)
(b, a)
(a, c)
(c, a)
(b, c)
(c, b)
(a, b, c)
(a, c, b)
(b, a, c)
(b, c, a)
(c, a, b)
(c, b, a)
```

If those all work in your branch (i.e. don't give Integral) and `integrate(s, *l).expand() == integrate(s.doit(), *l).expand()` then I guess everything is fine.

**Comment by smichr** [comment]
Just a couple of notes here:

* :repeat: this will need to be merged with an updated master which has changed and made this unmergable :eyeglasses: See **Synchronization with master sympy/sympy** at [the wiki](https://github.com/sympy/sympy/wiki/Development-workflow) 

* I might have just said this in another PR of yours, but (friendly reminder)
:information_source: could you please change the title of the PR? It's not very useful to have an issue number that can't be clicked there, and the number itself doesn't tell the reviewer what they are going to be seeing. (And when there are lots of Issue ### titles it's hard to remember what's what.)
:eyeglasses: See the section **Writing pull request title and description** at [the wiki](https://github.com/sympy/sympy/wiki/Development-workflow)

**Comment by kaushik94** [comment]
@smichr  is this okay ? But I see `>>>head` coming in the commit, what do I do ?

**Comment by smichr** [comment]
:bulb: From the [Zen of SymPy](https://github.com/sympy/sympy/wiki/The-Zen-of-SymPy): 
```
Smart tests are better than random tests.
But random tests sometimes find what your smartest test missed.
```

I don't know the answer, I am just raising a set of tests that I *think* should pass. Let me know if I am misunderstanding something, however:

```python
def test():
    from sympy import Sum, integrate
    from sympy.abc import a, b, c, x
    from sympy.utilities.iterables import permutations, subsets
    s = Sum(a + x, (x, b, c))
    failed = False
    for l in subsets((x, a, b, c)):
        if not l: continue
        for p in permutations(l):
            A = integrate(s, *l).expand()
            B = integrate(s.doit(), *l).expand()
            c1 = A == B
            c2 = A.doit().expand() == B.doit().expand()
            if c2 and not c1:
                print ('ok, but doit needed to be applied', p)
                failed = True
            elif not c2:
                print ('different results for',p)
                failed = True
            else:
                pass
    return not failed
```
Running gives
```python
>>> test()
('ok, but doit needed to be applied', (x,))
('ok, but doit needed to be applied', (a,))
('ok, but doit needed to be applied', (b,))
('ok, but doit needed to be applied', (c,))
('ok, but doit needed to be applied', (x, a))
('ok, but doit needed to be applied', (a, x))
('different results for', (a, b))
('different results for', (b, a))
('different results for', (a, c))
('different results for', (c, a))
('ok, but doit needed to be applied', (b, c))
('ok, but doit needed to be applied', (c, b))
('different results for', (a, b, c))
('different results for', (a, c, b))
('different results for', (b, a, c))
('different results for', (b, c, a))
('different results for', (c, a, b))
('different results for', (c, b, a))
False
```

**Comment by kaushik94** [comment]
@smichr  I feel there is a small mistake in your code but I do get what you were trying to say. The `doit` metho as been added. This was the result :
on my branch
```
('ok, but doit needed to be applied', (x,))
('ok, but doit needed to be applied', (a,))
('ok, but doit needed to be applied', (b,))
('ok, but doit needed to be applied', (c,))
('ok, but doit needed to be applied', (x, a))
('ok, but doit needed to be applied', (a, x))
('ok, but doit needed to be applied', (a, b))
('ok, but doit needed to be applied', (b, a))
('ok, but doit needed to be applied', (a, c))
('ok, but doit needed to be applied', (c, a))
('ok, but doit needed to be applied', (b, c))
('ok, but doit needed to be applied', (c, b))
('ok, but doit needed to be applied', (a, b, c))
('ok, but doit needed to be applied', (a, c, b))
('ok, but doit needed to be applied', (b, a, c))
('ok, but doit needed to be applied', (b, c, a))
('ok, but doit needed to be applied', (c, a, b))
('ok, but doit needed to be applied', (c, b, a))
```
and on master
```
('ok, but doit needed to be applied', (x,))
('ok, but doit needed to be applied', (a,))
('ok, but doit needed to be applied', (b,))
('ok, but doit needed to be applied', (c,))
('ok, but doit needed to be applied', (x, a))
('ok, but doit needed to be applied', (a, x))
('ok, but doit needed to be applied', (a, b))
('ok, but doit needed to be applied', (b, a))
('ok, but doit needed to be applied', (a, c))
('ok, but doit needed to be applied', (c, a))
('ok, but doit needed to be applied', (b, c))
('ok, but doit needed to be applied', (c, b))
('ok, but doit needed to be applied', (a, b, c))
('ok, but doit needed to be applied', (a, c, b))
('ok, but doit needed to be applied', (b, a, c))
('ok, but doit needed to be applied', (b, c, a))
('ok, but doit needed to be applied', (c, a, b))
('ok, but doit needed to be applied', (c, b, a))
```
The problem was that `s!=s.doit()` in your code. Not sure why, but i have added the doit method. Please review now.

**Comment by smichr** [comment]
:eyes: I would like someone else to look at this now. I don't know if something like the following should be evaluated any further or not

```python
>>> s
Sum(a + x, (x, b, c))
>>> integrate(s, b)
Integral(Sum(a + x, (x, b, c)), b)
```

**Comment by asmeurer** [comment]
You mean should it try to perform the summation to get the result?

File: sympy/integrals/integrals.py [hunk 2]
```diff
@@ -394,6 +395,12 @@ def doit(self, **hints):
             return S.Zero
 
         # now compute and check the function
+        if isinstance(self.function, Sum):
+            for variable in self.variables:
+                if variable in self.function.limits[0]:
```
Comment (smichr) [inline sympy/integrals/integrals.py]: limits[0] -> variables

Comment (smichr) [inline sympy/integrals/integrals.py]:
```python
>>> i=Integral(a,(b,1,2),c)
>>> i.limits
((b, 1, 2), (c,))
>>> i.variables
[b, c]
```

File: sympy/integrals/integrals.py [hunk 3]
```diff
@@ -394,6 +395,12 @@ def doit(self, **hints):
...
+                    break
+            else:
+                return Sum(Integral(self.function.function.doit(), self.variables[0]), self.function.limits).doit()
```
Comment (smichr) [inline sympy/integrals/integrals.py]: perhaps the following?
```
_i =self
_sum = self.function
return _sum.func(_i.func(_sum.function, *_i.limits), *_sum.limits)
```

We are responding to integrate(Sum(...)) so I suspect that leaving the Sum part undone (and the interior as well) is more in line with what the user wrote. If they want the Sum computed they have to apply doit or they should have done `integrate(summation(...)...)` in the first place.

File: sympy/integrals/integrals.py [hunk 4]
```diff
@@ -394,6 +395,12 @@ def doit(self, **hints):
...
         function = self.function
         if deep:
```
Comment (smichr) [inline sympy/integrals/integrals.py]: yes...the deep evaluation is handled if deep is True.

File: sympy/integrals/integrals.py (hunk 2 above)
Comment (kaushik94) [inline sympy/integrals/integrals.py]: But we are checking if `self`'s variables are in `self.function`'s limits right ?. Where is it going wrong ?

Comment (asmeurer) [inline sympy/integrals/integrals.py]: This only checks the first limit.

**Comment by kaushik94** [comment]
I think this has been fixed

**Comment by tyagi-prashant** [comment]
@asmeurer what is wrong with above implementation, i worked on this issue and come up with same solution that kaushik94 does.

**Comment by asmeurer** [comment]
Looking at the code I think the only issue is that it doesn't handle multiple Sums. I'm not sure why it was closed, though.

**Comment by czgdp1807** [comment]
@kaushik94 Would you still like to continue work on it? The issue is still un-resolved. 
Thanks for the contributions.

**Comment by kaushik94** [comment]
> @kaushik94 Would you still like to continue work on it? The issue is still un-resolved.
> Thanks for the contributions.

Hey @czgdp1807 yes I would like to work on this issue. Please allow me a couple of days to update the codebase and understand the review in more detail. Thank you so much for the help.

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/sympy/sympy/pull/18491

**Pull request closed without merging**